include/zephyr/net/tls_credentials.h

```diff
@@ -124,6 +124,37 @@ int tls_credential_get(sec_tag_t tag, enum tls_credential_type type,
  */
 int tls_credential_delete(sec_tag_t tag, enum tls_credential_type type);
 
+/**
+ * @brief Create client private key and certificate signing request (CSR).
+ *
+ * @details This function generates a public/private keypair, and then signs a corresponding PKCS#10
```

Comment:
add that `SECP256R1` is used for key generation.

or add a config struct to allow user to specify key type/size along with CSR info such as country, organization, etc.  (instead of only the `char *dn` parameter).

Reply:
> add that SECP256R1 is used for key generation.

The thought I had was that it could simply be up to the storage back-end which key type is used for generating the CSR.

> or add a config struct to allow user to specify key type/size along with CSR info such as country, organization, etc. (instead of only the char *dn parameter).

The `dn` parameter does allow a full distinguished name to be specified, so country, organization, etc are all possible. I'd rather avoid re-inventing a representation for those structs when MbedTLS already accepts them quite naturally as a string.

As for key-type customization, I am struggling to see a way to do this cleanly. There are a lot of concerns to balance:

1. If we need key type customization, we probably need it in command line too, so the system has to be compatible with that
2. We should, IMO, try to keep the tls_credentials module API agnostic of the cryptography backend being used (such as MbedTLS). So we cannot simply wrap MbedTLS types (very easily)
3. Others have expressed interest in expanding the TLS Credentials module to allow multiple simultaneous backends / APIs to create additional backends.

I suspect the cleanest way to handle this would probably be to allow a custom keytype to be requested via an optional string argument (alongside the dn argument). It would be up to the back-end to interpret/parse the keytype string, but this is probably the most versatile / interoperable option possible. I'd be willing to implement that in this PR.

Either that, or we will need to develop a full backend options customization system (similar to what sockets provide), and that, I think, should be a separate PR.

What do you think of the strings solution? @rlubos What are your thoughts on all of this as well?

Reply:
> The dn parameter does allow a full distinguished name to be specified, so country, organization, etc are all possible.

Ah ok, I didn't realize that accepted everything 👍🏻

Reply:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/4138391/a2bd5e02-de2b-4a5e-b58b-9f2c53596c16)

I am referring to this.
Either indicate in the function header what it does or allow the user to specify their own key parameters.

Reply:
Yeah, I understood you -- That section of code you highlight is in the secure credentials backend. While this is currently the only backend which supports CSR generation, it is conceivable that there might be other backends which also support CSR generation eventually.

Therefore, if we are letting the backends decide what type of key to generate, it doesn't make sense to specify that in the header file (which describes the abstract API all backends conform to)

Allowing users to specify a desired key type for the back-end to consider does make some sense, But if we're going to do it, I think it will either need to come as part of another PR, or be a simple string argument which the backend is responsible for interpreting

Reply:
I'm only asking that you pick one option:
Either document that only ecc 256 keys are supported or allow the user to specify.

Maybe the documenting should be done in kconfig, maybe with an option?

Reply:
> Either document that only ecc 256 keys are supported 

I could document it in the back-end for which it is true, sure.

But I cannot document it in `tls_credentials.h` since which keyt-ypes are supported depends on the back-end in use.

> or allow the user to specify.

The problem with that is that it is complicated to do cleanly. I was asking your opinion on using a pure string to allow this.

But I have decided after some research that a simple enum listing a subset of the IANA supported groups is probably sufficient for this purpose.

Reply:
> But I cannot document it in tls_credentials.h since which key-types are supported depends on the back-end in use.

i understand, i suggested using kconfig.

> a simple enum listing a subset of the IANA supported groups is probably sufficient for this purpose.

sounds good.

Reply:
I'm not really an expert in the area of key generation, but from the API perspective I think it'd make sense to provide an option to select key type (or parameters?) instead of hardcoding this within the backend. If the backed does not support requested params, it could simply reject the request.